game/scripts/vscripts/Sections/Chapter4/SectionWards.ts

```diff
@@ -0,0 +1,178 @@
+    const isGoalStarted = (key: NeutralGoalKeys) =>
+        context[key] === GoalState.Started ||
+        context[key] === GoalState.Completed;
+    const isGoalCompleted = (key: NeutralGoalKeys) =>
+        context[key] === GoalState.Completed;
+
+    const goals: Goal[] = [];
+    const addGoal = (key: NeutralGoalKeys, text: string) => {
+        if (isGoalStarted(key)) {
+            goals.push({ text: text, completed: isGoalCompleted(key) });
+        }
+    };
+
+    addGoal(NeutralGoalKeys.FetchWards, "Go pick those up and come back here.");
+    addGoal(NeutralGoalKeys.PlaceObserverWard, "Lets put an observer ward on this high ground.");
+    addGoal(NeutralGoalKeys.PlaceSentryWard, "Lets put a sentry ward on this high ground.");
+    return goals;
+};
+
+const invisHeroInfo = [
+    { name: "npc_dota_hero_riki", loc: GetGroundPosition(Vector(500, -4200), undefined) },
+    { name: "npc_dota_hero_mirana", loc: GetGroundPosition(Vector(500, -4000), undefined) },
+    { name: "npc_dota_hero_clinkz", loc: GetGroundPosition(Vector(700, -3900), undefined) },
+    { name: "npc_dota_hero_bounty_hunter", loc: GetGroundPosition(Vector(900, -3700), undefined) },
+    { name: "npc_dota_hero_invoker", loc: GetGroundPosition(Vector(600, -4800), undefined) },
+    { name: "npc_dota_hero_nyx_assassin", loc: GetGroundPosition(Vector(600, -4500), undefined) },
+    { name: "npc_dota_hero_slark", loc: GetGroundPosition(Vector(900, -4500), undefined) },
+    { name: "npc_dota_hero_weaver", loc: GetGroundPosition(Vector(400, -4500), undefined) },
+    { name: "npc_dota_hero_sand_king", loc: GetGroundPosition(Vector(400, -4800), undefined) },
+];
+
+function onStart(complete: () => void) {
+    print("Starting", sectionName);
+
+    const playerHero = getOrError(getPlayerHero(), "Could not find the player's hero.");
+    playerHero.SetMoveCapability(UnitMoveCapability.GROUND);
+
+    const observerWardItem = CreateItem("item_ward_observer", undefined, undefined);
+    const sentryWardItem = CreateItem("item_ward_sentry", undefined, undefined);
+
+    graph = tg.forkAny([
+        tg.trackGoals(getGoals),
+        tg.seq([
+            tg.fork(invisHeroInfo.map(hero => tg.spawnUnit(hero.name, hero.loc, DotaTeam.BADGUYS, hero.name))),
+            tg.immediate(ctx => {
+                for (const invisHero of invisHeroInfo) {
+                    const hero: CDOTA_BaseNPC_Hero = ctx[invisHero.name];
+                    hero.AddNewModifier(undefined, undefined, "modifier_invisible", undefined);
+                }
+            }),
+
+            tg.wait(1),
+
+            tg.immediate(() => {
+                CreateItemOnPositionSync(Vector(2200, -3800), observerWardItem)
+                CreateItemOnPositionSync(Vector(2200, -3900), sentryWardItem)
+            }),
+
+            tg.immediate(context => context[NeutralGoalKeys.FetchWards] = GoalState.Started),
+            tg.completeOnCheck(() => playerHero.HasItemInInventory("item_ward_dispenser"), 1),
```

Comment:
This only checks if the player has picked up one of the two wards right? Couldn't we somehow check if both are picked up?

Reply:
item_ward_dispenser is the combined one (obs & sentry), so they need to pick up both of them before this completes